rules/how-to-read-source-code/rule.md

```diff
@@ -0,0 +1,40 @@
+  - title: Matt Goldman
+    url: https://www.ssw.com.au/people/matt-goldman
+created: 2022-09-12T14:44:28.666Z
+guid: 1b706f91-a99c-49a6-9ae8-72d2026abc11
+---
+First of all, you need to have the following prerequisites so that you can read the code smoothly afterwards.
+
+1. **Basic knowledge** - Knowledge of relevant languages ​​and underlying technologies.
+2. **Software function** - You must first know what kind of functions this software completes, what features it has, and what configurations it has. You need to read the user manual first, then let the software run, and feel it for yourself.
+3. **Relevant documentation** - Read the relevant internal documents, Readme or Release Notes, Design or Wiki. These documents can let you understand all aspects of the software. If your software doesn't have documentation, then you can only count on the original author of the software still alive and willing to communicate.
+4. **The structure of the code** - You need to know what is the function of each directory. If the program you want to read is organized under some standard framework, such as the [Clean Architecture](https://github.com/jasontaylordev/CleanArchitecture). Then congratulations, the code is not difficult to read.
+
+Next, you need to understand what parts of the code of this software are made up of. Below is a list for reference.
+
+1. **Interface/abstract definition** - Any code will have many interface or abstract definitions, which describe the data structures or business entities that the code needs to deal with and the relationships between them. It is very important to understand these relationships.
+2. **Module adhesive layer** - A lot of our code is used to glue code, such as middleware, Promises pattern, Callback, proxy and delegation, dependency injection and so on. The gluing techniques between these code modules are very important. Because they will split the code that would otherwise be straightforward, making it difficult for you to understand their relationships.
+3. **Business Process** - This is how the code runs. In the beginning, we don't want to go into the details. But we need to figure out at a high level what the entire business process looks like. And in this process, how data is passed and processed. Generally speaking, we need to draw program flow charts.
+4. **Detailed implementation** - After understanding the above three aspects, you will have a general understanding of the framework and logic of the entire code. At this point, you can dive into the details and start reading the code for the specific implementation. In general, you need to know the following facts, which will help you find the key points when reading the code.
+    - **Code logic** - The code has two kinds of logic. One is business logic. The other is control logic. You need to separate these two kinds of logic. The reason why many codes are confusing is that these two kinds of logic are mixed.
```

Comment:
change:
> codes

to:
code bases

Reply:
In a few places (so won't call out each one).

Reply:
> In a few places (so won't call out each one).

For other places, I changed `codes` to `code`, `this kind of code`, `they` since [code is a mass noun in computer science](https://english.stackexchange.com/questions/20455/is-it-wrong-to-use-the-word-codes-in-a-programming-context) and it's not referring the code base.